src_c/math.c

```diff
@@ -792,6 +796,75 @@ vector_copy(pgVector *self)
     return (PyObject *)ret;
 }
 
+static PyObject *
+vector_clamp_magnitude(pgVector *self, PyObject *args, PyObject *kwargs)
+{
+    Py_ssize_t i;
+    pgVector *ret;
+
+    ret = (pgVector *)pgVector_NEW(self->dim);
+    if (ret == NULL)
+        return NULL;
+
+    for (i = 0; i < self->dim; ++i)
+        ret->coords[i] = self->coords[i];
+
+    vector_clamp_magnitude_ip(ret, args, kwargs);
+
+    return (PyObject *)ret;
+}
+
+static PyObject *
+vector_clamp_magnitude_ip(pgVector *self, PyObject *args, PyObject *kwargs)
+{
+    Py_ssize_t i;
+    double min_length = 0; /* Default minimum. */
+    double max_length;
+    double old_length;
+    double fraction;
+    static char *keywords[] = {"max_length", "min_length", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "d|d", keywords,
+                                     &max_length, &min_length)) {
+        return NULL;
+    }
+
+    /* Get magnitude of Vector */
+    old_length = sqrt(_scalar_product(self->coords, self->coords, self->dim));
+
+    if (old_length == 0) {
+        PyErr_SetString(PyExc_ValueError,
+                        "Cannot clamp a vector with zero length");
+        return NULL;
+    }
+
+    if (old_length > max_length) {
+        /* Scale to length */
+        old_length =
+            sqrt(_scalar_product(self->coords, self->coords, self->dim));
+
+        fraction = max_length / old_length;
+        for (i = 0; i < self->dim; ++i)
+            self->coords[i] *= fraction;
+
+        Py_RETURN_NONE;
+    }
+
+    else if (old_length < min_length) {
+        /* Scale to length */
+        old_length =
+            sqrt(_scalar_product(self->coords, self->coords, self->dim));
+
+        fraction = min_length / old_length;
+        for (i = 0; i < self->dim; ++i)
+            self->coords[i] *= fraction;
+
+        Py_RETURN_NONE;
```

Comment:
this looks like the same code as in the other if block... maybe this can be extracted so the code duplication is removed?

Reply:
There's a difference between the two since in the variable `fraction`, the numerator is different in both blocks of code. Since C doesn't support nested functions, I'm not sure how I'm able to remove this duplicate code.